Ddc/avro config fix

btnewton opened the pull request:
Fixes a bug that prevented Avro data configuration from being used. User configurations were not used so properties were using their respective default values. 

This effected:
 
- enhanced.avro.schema.support (defaults to false)
- connect.meta.data (defaults to true)
- schemas.cache.config (customizable but was being set using the subtly different schema.cache.size  property - that behavior is maintained here: https://github.com/btnewton/kafka-connect-storage-cloud/blob/8a6fdcba5cc98da60cb3e23c3abd2c4fe71325f1/kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/avro/AvroFormat.java#L40)

ConfluentJenkins commented:
Can one of the admins verify this patch?

btnewton commented:
@ConfluentCLABot [clabot:check]

kkonstantine commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/avro/AvroDataConfig.java:
```diff
@@ -0,0 +1,72 @@
+ * Adapts {@link io.confluent.connect.avro.AvroDataConfig} which uses {@link io.confluent.common.config.AbstractConfig} which is
+ * incompatible with {@link org.apache.kafka.common.config.AbstractConfig}.
+ */
+public class AvroDataConfig extends AbstractConfig implements ComposableConfig {
```
It's indeed unfortunate that we can't inherit from `AvroDataConfig` in schema-registry. It would be more bullet proof to future changes/additions to this config class (that's how we missed passing this functionality here in the first place. It was added in SR but not in the sink connectors). 

Still, since this is going to be used by the hdfs connector as well, I'll take this part and issue a PR in kafka-connect-storage-common instead.

kkonstantine replied:
I'll also call it `ComposableAvroDataConfig`. Maybe not a great name, but will help with conflicts with the original type of `AvroDataConfig` which we'll need to use when accessing the static members of this class.

kkonstantine commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/avro/AvroFormat.java:
```diff
@@ -17,20 +17,30 @@
-    this.avroData = new AvroData(storage.conf().getInt(S3SinkConnectorConfig.SCHEMA_CACHE_SIZE_CONFIG));
+    Map<String, Object> avroDataProps = io.confluent.connect.avro.AvroDataConfig.baseConfigDef()
+        .parse(storage.conf().plainValues());
+    avroDataProps.put(SCHEMAS_CACHE_SIZE_CONFIG, storage.conf()
```
Since recently we use one argument per line in multi-line method calls.

kkonstantine commented on kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/avro/AvroFormat.java:
```diff
@@ -17,20 +17,30 @@
   public AvroFormat(S3Storage storage) {
     this.storage = storage;
-    this.avroData = new AvroData(storage.conf().getInt(S3SinkConnectorConfig.SCHEMA_CACHE_SIZE_CONFIG));
+    Map<String, Object> avroDataProps = io.confluent.connect.avro.AvroDataConfig.baseConfigDef()
```
I feel that this code block here (parsing and overwriting the `AvroDataConfig` values) rather belongs in an `avroDataConfig` method within `S3SinkConnectorConfig` and not here.

kkonstantine commented on kafka-connect-s3/src/test/java/io/confluent/connect/s3/S3SinkConnectorConfigTest.java:
```diff
@@ -102,6 +103,15 @@ public void testRecommendedValues() throws Exception {
 
   @Test
+  public void testAvroDataConfigSupported() throws Exception {
+    properties.put(AvroDataConfig.ENHANCED_AVRO_SCHEMA_SUPPORT_CONFIG, "true");
```
This test could actually test the `avroDataConfig` method mentioned above, and make sure the override of schema cache size takes place too.

btnewton commented:
@kkonstantine Thanks for taking the time to review my PR!

I've moved the AvroDataConfig setup to a S3ConnectorConfig method and updated the unit test. I'm ready to pull in your ComposableAvroDataConfig change as soon as its ready. Let me know if you need anything else changed.

btnewton commented:
@kkonstantine I've updated this PR to make use of the `ComposableAvroDataConfig` class you suggested. `ComposableAvroDataConfig` has been implemented in [PR 42](https://github.com/confluentinc/kafka-connect-storage-common/pull/42) of kafka-connect-storage-common. Please let me know if you need anything else.

kkonstantine commented:
So, I've made a few necessary changes, starting with your PR here as a basis, to incorporate the changes I mention here: https://github.com/confluentinc/kafka-connect-storage-common/pull/45 and enable the configs finally in the connector. 

Would you like me to attempt and push the changes on top of your PR here (overriding some code) or just issue another separate PR? I'm fine either way.

btnewton commented:
Hi @kkonstantine yes, please add your changes in whatever way is easiest. Thanks for your help

kkonstantine commented:
@btnewton thanks for your contribution! I ended up issuing a new PR with your initial changes and my additions here: https://github.com/confluentinc/kafka-connect-storage-cloud/pull/132

Given that this was merged and will appear in the next bugfix release for `4.0.x` I'm closing this PR as well. Cheers!
